ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Follower.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Consensus.HardFork.History.Follower (
+    -- * API
+    HistoryFollower (..)
+  , WithInterpreter (..)
+  , WithInterDefault
+  , mkWithInterpreter
+  , newHistoryFollower
+    -- * Client
+  , ClientHistoryFollower (..)
+  , WithInterClient
+  , newClientFollower
+  ) where
+
+import           Control.Monad.Class.MonadSTM.Strict hiding (newTVarIO)
+import           Data.Either (isRight)
+import           Data.SOP.Strict
+
+import           Ouroboros.Consensus.HardFork.Abstract
+import           Ouroboros.Consensus.HardFork.Combinator.Basics
+import           Ouroboros.Consensus.HardFork.Combinator.Ledger ()
+import qualified Ouroboros.Consensus.HardFork.History as History
+import           Ouroboros.Consensus.HardFork.History.Qry
+import           Ouroboros.Consensus.Ledger.Basics
+import           Ouroboros.Consensus.Ledger.Extended
+import           Ouroboros.Consensus.Storage.ChainDB
+import           Ouroboros.Consensus.Util
+import           Ouroboros.Consensus.Util.IOLike
+
+import           Ouroboros.Network.Block
+import           Ouroboros.Network.Point
+
+
+-- | The 'HistoryFollower' can be used to extend the 'Follower' with metadata
+-- which come from the ledger state. Currently this metadata is a History
+-- 'Interpreter' but it's technically possible to generalise this.
+--
+-- In the same way that a 'Follower' keeps track of the tip of a user, the
+-- 'HistoryFollower' keeps track and mirrors the Interpreter that a user has.
+-- Everytime a new 'Interpreter' is created and returned to the user, it is also
+-- cached. The api returns 'Nothing' when the cached 'Interpreter' is capable
+-- of serving the specified point, so the user doesn't need a new one.
+--
+newtype HistoryFollower m blk = HistoryFollower {
+      getInterpreter :: Point blk -> STM m (Maybe (Interpreter (HardForkIndices blk)))
+    }
+
+data WithInterpreter blk b = WithInterpreter {
+      payload     :: b
+    , interpreter :: Maybe (Interpreter (HardForkIndices blk))
+    } deriving Show
+
+type WithInterDefault blk = WithInterpreter blk (Serialised blk)
```

Comment:
Is this "default" as in "not a client"? If so, I'd suggested `WithInterSerialised` or `WithInterNTN` (for node-to-node).

Also, more weakly, I'd suggesting spelling out `Inter` and `Interpreter`; `Inter` seems a bit ambiguous at first blush.

Reply:
> Is this "default" as in "not a client"? If so, I'd suggested WithInterSerialised or WithInterNTN (for node-to-node -- and WithInterClient would probably then best be WithInterNTC).

I used `WithInterDefault` following the existing naming `DefaultCodecs`, which is used by the server and contains the `Serialized` block. However it could make sence to change `Default` to `Server` in all these places. Note that the Interpreter is only involved with the NodeToClient protocol.

> Also, more weakly, I'd suggesting spelling out Inter as Interpreter; Inter seems a bit ambiguous at first blush.

In general I used the shorter `Inter` when the name started to become too long. But I'm fine with your suggestion.